dlls/nvcuda/nvcuda.c

```diff
@@ -3651,23 +3652,122 @@ CUresult WINAPI wine_cuDeviceGetUuid(CUuuid *uuid, CUdevice dev)
+}
+
+static ULONG query_reg_value( HKEY hkey, const WCHAR *name,
+                       KEY_VALUE_PARTIAL_INFORMATION *info, ULONG size )
+{
+    unsigned int name_size = name ? lstrlenW( name ) * sizeof(WCHAR) : 0;
+    UNICODE_STRING nameW = { name_size, name_size, (WCHAR *)name };
+
+    if (NtQueryValueKey( hkey, &nameW, KeyValuePartialInformation,
+                         info, size, &size ))
+        return 0;
+
+    return size - FIELD_OFFSET(KEY_VALUE_PARTIAL_INFORMATION, Data);
+}
+
+/* code borrowed from Wine ends here */
+
 CUresult WINAPI wine_cuDeviceGetLuid(char *luid, unsigned int *deviceNodeMask, CUdevice dev)
 {
-    int wine_luid[] = { 0x0000adde, 0x00000000 };
+    char buffer[1024];
+    KEY_VALUE_PARTIAL_INFORMATION *value = (void *)buffer;
+    KEY_BASIC_INFORMATION *key = (void *)buffer;
+    HKEY pci, outerdev, innerdev, prop;
+    DWORD size, i = 0;
+    CUresult result;
+    CUuuid uuid;
 
     TRACE("(%p, %p, %d)\n", luid, deviceNodeMask, dev);
-    /* Linux native libcuda does not provide a LUID, so we need to fake something and return a success */
 
-    memcpy(luid, &wine_luid, sizeof(wine_luid));
-    FIXME("Fix this LUID: ");
-    for(int i = 0; i < sizeof(luid); i++){
-        FIXME("%02x", (unsigned char)luid[i]);
-        if(i == 3) FIXME("-");
+    if ((result = pcuDeviceGetUuid(&uuid, dev)) != CUDA_SUCCESS)
+        return result;
+
+    result = CUDA_ERROR_NOT_SUPPORTED;
+
+    if (!(pci = reg_open_ascii_key(NULL, "\\Registry\\Machine\\System\\CurrentControlSet\\Enum\\PCI")))
+        return result;
+
+    while (!NtEnumerateKey(pci, i++, KeyBasicInformation, key, sizeof(buffer), &size))
+    {
+        DWORD j = 0;
+
+        if (!(outerdev = reg_open_key(pci, key->Name, key->NameLength)))
+            continue;
+
+        while (!NtEnumerateKey(outerdev, j++, KeyBasicInformation, key, sizeof(buffer), &size))
+        {
+            if (!(innerdev = reg_open_key(outerdev, key->Name, key->NameLength)))
+                continue;
+
+            if (!(prop = reg_open_ascii_key(innerdev, devpropkey_gpu_vulkan_uuidA)))
+                goto next;
+
+            if (query_reg_value(prop, NULL, value, sizeof(buffer)) != sizeof(CUuuid) || memcmp(value->Data, &uuid, sizeof(uuid)))
+                goto cleanup;
+
+            NtClose(prop);
+
+            if (!(prop = reg_open_ascii_key(innerdev, devpropkey_gpu_luidA)))
+                goto next;
+
+            if (query_reg_value(prop, NULL, value, sizeof(buffer)) != sizeof(LUID))
+                goto cleanup;
+
+            memcpy(luid, value->Data, sizeof(LUID));
+            *deviceNodeMask = 1;
+            result = CUDA_SUCCESS;
+
+            TRACE("Found LUID: %02x%02x%02x%02x-%02x%02x%02x%02x\n",
+                (unsigned char)luid[0], (unsigned char)luid[1], (unsigned char)luid[2], (unsigned char)luid[3],
+                (unsigned char)luid[4], (unsigned char)luid[5], (unsigned char)luid[6], (unsigned char)luid[7]);
```

Comment:
I kind of like this one: 
```
    TRACE("Found LUID: ");
    for(int i = 0; i < sizeof(luid); i++){
        TRACE("%02x", (unsigned char)luid[i]);
        if(i == 3) TRACE("-");
```
for the printout. Then again.. might be more of an overhead doing a for-loop instead of printing directly like you do here... Open for suggestions.

Reply:
To be honest, both versions are horrible :sweat_smile:

But myself I prefer a single "call" (well, function-like macro) for fixed-size loops. I considered just removing it but I wasn't sure if registry lookup wasn't too brittle, at least this way we'll know if it succeeded, and it it matches LUID as known from other APIs or not.

Reply:
Or we could move that nonsense into some kind of `wine_dbgstr_culuid`, similar to [`wine_dbgstr_guid`](https://github.com/SveSop/nvcuda/blob/master/include/wine/debug.h#L322-L330), I'm just not sure where should that function live.